velox/exec/Window.cpp

```diff
@@ -0,0 +1,487 @@
+    allFuncsFrameEndBuffer_.push_back(frameEndBuffer);
+
+    auto rawFrameStartBuffer = frameStartBuffer->asMutable<vector_size_t>();
+    auto rawFrameEndBuffer = frameEndBuffer->asMutable<vector_size_t>();
+    allFuncsRawFrameStartBuffer_.push_back(rawFrameStartBuffer);
+    allFuncsRawFrameEndBuffer_.push_back(rawFrameEndBuffer);
+  }
+}
+
+// This function computes the auxiliary structure partitionStartRows_
+// which stores the starting row for each input partition in the data.
+void Window::computePartitionStartRows() {
+  // Randomly assuming that max 10000 partitions are in the data.
+  partitionStartRows_.reserve(10000);
+  auto partitionCompare = [&](const char* lhs, const char* rhs) -> bool {
+    return compareRowsWithKeys(lhs, rhs, partitionKeyInfo_);
+  };
+
+  // Using a sequential traversal to find changing partitions.
+  // This algorithm is inefficient and can be changed
+  // i) Use a binary search kind of strategy.
+  // ii) If we use a Hashtable instead of a full sort then the count
+  // of rows in the partition can be directly used.
+  partitionStartRows_.push_back(0);
+
+  VELOX_CHECK_GT(sortedRows_.size(), 0);
+  for (auto i = 1; i < sortedRows_.size(); i++) {
+    if (partitionCompare(sortedRows_[i - 1], sortedRows_[i])) {
+      partitionStartRows_.push_back(i);
+    }
+  }
+
+  // Setting the startRow of the (last + 1) partition to be returningRows.size()
+  // to help for last partition related calculations.
+  partitionStartRows_.push_back(sortedRows_.size());
+}
+
+// This function orders the input rows by (partition keys + order by keys).
+// Doing so orders all rows of a partition adjacent to each other
+// and sorted by the ORDER BY clause. This is the order in which the rows
+// will be output by this operator.
+void Window::sortPartitions() {
+  // This is a very inefficient but easy implementation to order the input rows
+  // by partition keys + sort keys.
+  // Sort the pointers to the rows in RowContainer (data_) instead of sorting
+  // the rows.
+  sortedRows_.resize(numRows_);
+  RowContainerIterator iter;
+  data_->listRows(&iter, numRows_, sortedRows_.data());
+
+  std::sort(
+      sortedRows_.begin(),
+      sortedRows_.end(),
+      [this](const char* leftRow, const char* rightRow) {
+        return compareRowsWithKeys(leftRow, rightRow, allKeyInfo_);
+      });
+
+  computePartitionStartRows();
+
+  currentPartition_ = 0;
+}
+
+void Window::noMoreInput() {
+  Operator::noMoreInput();
+  // No data.
+  if (numRows_ == 0) {
+    finished_ = true;
+    return;
+  }
+
+  // At this point we have seen all the input rows. We can start
+  // outputting rows now.
+  // However, some preparation is needed. The rows should be
+  // separated into partitions and sort by ORDER BY keys within
+  // the partition. This will order the rows for getOutput().
+  sortPartitions();
+  createPeerAndFrameBuffers();
+}
+
+// This function is to find the frame end points for the current row
```

Comment:
Move this documentation to the header file. Drop ":" between parameter name and description.

Reply:
I had left this documentation for future writing of this code. But moving to the header file is fine also. Have moved it.